File: Server_trunk/res/scripts/cell/Spell/Effects/EffectHitPose.py
# -*- coding: utf-8 -*-
#
"""
"""
from KBEDebug import *
from ..Spell import SpellEffect

class EffectHitPose(SpellEffect):
	"""
	受击效果：击中、击退、击飞、击倒等
	"""
	def init(self, dataSection):
		"""
		"""
		SpellEffect.init(self, dataSection)
		
		# 标识技能和声音会在对象的哪个子对象上播放，例如：Bip01/r_h/s014_lod0/WeaponTrail
		#self.bindingObjectPath = dataSection.readString( "bindingObjectPath" )
		
		# 击中的类型：详看eHitPose定义
		#self.hitPoseType = dataSection.readString( "hitPoseType" )
		
		# 击中的光效
		#self.effect = dataSection.readString( "effect" )
		
		# 击中的声音
		#self.sound = dataSection.readString( "sound" )
	
	def cast(self, src, dst, spell):
		"""
		src向dst施放一个效果
		@param src: Entity; 施法者
		@param dst: Entity; 受术者
		@param spell: instance of Spell; 表示是由哪个技能触发的效果
		@return: None
		"""
		# 允许没有src
		srcID = -1
		if src is not None:
			srcID = src.id

		dst.allClients.seeSpellEffectFS( self.id, srcID )

File: Client_trunk2/Assets/Scripts/Spells/Effects/EffectHitPose.cs
using UnityEngine;
using System.Collections;

namespace SPELL
{
	/// <summary>
	/// 受击动作组件
	/// </summary>
	public class EffectHitPose : SpellEffect
	{
		//public static string[] s_animationName = { "hit_normal", "hit_up", "hit_down", "hit_back" };

		public string bindingObjectPath = "";        // 标识技能和声音会在对象的哪个子对象上播放，例如：Bip01/r_h/s014_lod0/WeaponTrail
        public int hitPoseType;                      // animation name which definition in Animator.
		public string effect = "";                   // effect name with prefab
		public string sound = "";                    // sound name

		public override void Init(DataSection.DataSection dataSection)
		{
			base.Init(dataSection);
            hitPoseType = dataSection.readInt("param1");
            bindingObjectPath = dataSection.readString("param2");
            effect = dataSection.readString("param3");
            sound = dataSection.readString("param4");
		}

		/// <summary>
		/// caster do something to dst use by spell
		/// </summary>
		/// <param name="spell">Spell.</param>
		/// <param name="src">spell caster.</param>
		/// <param name="dst">spell receiver.</param>
		public override void Cast(AvatarComponent src, AvatarComponent dst, SpellEx spell, SpellTargetData targetData)
		{

			Animator animator = dst.animator;
            if (animator)
            {
                animator.SetInteger("hit_pose", hitPoseType);
                animator.SetTrigger("hit");
            }

			Transform transf = dst.myTransform;
			if (bindingObjectPath != "")
			{
				transf = transf.FindChild(bindingObjectPath);
				if (!transf)
					transf = dst.myTransform;  // set to src position if not found
			}

			// 播放光效
			if (effect != "")
                dst.effectManager.AddEffect(effect, transf.position);

            // 播放声音
            if (sound != "")
				SoundFXManager.Instance.PlaySound(sound, transf);
		}
		
	}
}
